lib/cloud_controller/copilot/adapter.rb

```diff
@@ -54,9 +54,16 @@ def delete_capi_diego_process_association(process)
 
         def bulk_sync(routes:, route_mappings:, processes:)
           copilot_client.bulk_sync(
-            routes: routes.map { |r| { guid: r.guid, host: r.fqdn, path: r.path } },
-            route_mappings: route_mappings.map { |rm| { capi_process_guid: rm.process.guid, route_guid: rm.route.guid, route_weight: rm.weight } },
-            capi_diego_process_associations: processes.map do |process|
+            routes: routes.compact.map { |r| { guid: r.guid, host: r.fqdn, path: r.path } },
+            route_mappings: route_mappings.compact.map do |rm|
+              next if rm.process.nil? || rm.route.nil?
+              {
+                capi_process_guid: rm.process.guid,
```

Comment:
Calling `rm.process` and `rm.route` actually does a database query under the hood. Since we're fetching all of these resources upfront in the calling class, we're going to be making **a lot** of queries against ccdb here in large environments ([see n+1 queries problem](https://semaphoreci.com/blog/2017/08/09/faster-rails-eliminating-n-plus-one-queries.html)).

If we only need the guids you could do something like `rm.process_guid`, but I'm not sure that will work with your existence check above.

I think a better solution would be to do this work up front in the calling class if possible. We should also implement some form of pagination there as well (might be out of scope for this part, but it needs to happen before this could go on a production environment)... 🙂